src/libopenrave/kinbodystatesaver.cpp

```diff
@@ -18,6 +18,121 @@
 
 namespace OpenRAVE {
 
+/// \brief push link to _listNonCollidingLinksWhenGrabbed of grabbed.
+static void _PushLinkToListNonCollidingLinksWhenGrabbed(Grabbed& grabbed,
+                                                        const int linkindex,
+                                                        const KinBody::LinkConstPtr& pLinkRef,
+                                                        const std::vector<KinBody::LinkPtr>& vLinks,
+                                                        const EnvironmentBasePtr& pEnv)
+{
+    if( linkindex < 0 || linkindex >= (int)vLinks.size() ) {
+        RAVELOG_WARN_FORMAT("env=%s, could not restore link '%s' since its index %d is out of range (body num links is %d)", pEnv->GetNameId()%pLinkRef->GetName()%pLinkRef->GetIndex()%vLinks.size());
+    }
+    else {
+        grabbed._listNonCollidingLinksWhenGrabbed.push_back(vLinks.at(linkindex));
+    }
+}
+
+void KinBody::_RestoreGrabbedBodiesFromSavedData(const KinBody& savedBody,
+                                                 const int options,
+                                                 const std::unordered_map<int, GrabbedPtr>& savedGrabbedBodiesByEnvironmentIndex)
+{
+    {
+        // have to release all grabbed first
+        ReleaseAllGrabbed();
+        OPENRAVE_ASSERT_OP(_grabbedBodiesByEnvironmentIndex.size(),==,0);
+        for (const std::unordered_map<int, GrabbedPtr>::value_type& grabPair : savedGrabbedBodiesByEnvironmentIndex) {
```

Comment:
You use use `MapGrabbedByEnvironmentIndex::value_type` to ensure that if the type changes we don't have to update it everywhere

Reply:
@rschlaikjer 

Oh, the original code is written by you https://github.com/rdiankov/openrave/commit/9176271a500f12fbb3f1392fd839a329c40f9189.

So, do you want to replace this by `MapGrabbedByEnvironmentIndex::value_type`, and to  slightly re-order the declaration of `MapGrabbedByEnvironmentIndex` so that we can replace the same thing in `KinBodyStateSaver` member?

Even though, can we do that change in another PR, since it's touching the code line to update the saved data contents?
https://github.com/rdiankov/openrave/pull/1430

Thanks!

Reply:
Oh I see, this is outside the range of that using-decl. It was a small comment, feel free to leave it or extract that decl in a different PR.

Reply:
thanks!
I'll try revisiting in the next PR!